test/e2e/network/topology.go

```diff
@@ -0,0 +1,205 @@
+		}()
+		err = jig.CheckServiceReachability(svc, execPod)
+		framework.ExpectNoError(err)
+
+		// poll the service 100 times and get a per zone traffic distribution
+		zones := map[string]int{}
+		maxConnections := 100
+		nc := fmt.Sprintf(`echo hostName | nc -v -w 5 %s %d`, svcIP, port)
+		cmd := fmt.Sprintf("for i in $(seq 0 %d); do echo; %s ; done", maxConnections, nc)
+		stdout, err := framework.RunHostCmd(execPod.Namespace, execPod.Name, cmd)
+		framework.ExpectNoError(err)
+
+		hostnames := strings.Split(stdout, "\n")
+		failed := 0
+		// failed request return an empty string that we can not associate to a zone
+		for _, h := range hostnames {
+			z, ok := podsZones[h]
+			if ok {
+				zones[z]++
+			} else {
+				failed++
+			}
+		}
+		framework.Logf("Connections from Zone %s distributed with topology hints enabled %v Failed: %d", nodeZone, zones, failed)
+		// Fail if the traffic in the zone is not higher than the 65%
+		// don't consider the failed connections for the stats
+		// TODO(aojea): based on local testing with KIND revisit percentages.
+		zoneTraffic := zones[nodeZone] * 100 / (maxConnections - failed)
+		if zoneTraffic < 65 {
+			framework.Failf("Traffic within the zone is lower than 65 per cent : %d", zoneTraffic)
+		}
+
+		// disable topology aware hints on the service
+		ginkgo.By("disabled topology aware annotation on service")
+		svc, err = jig.UpdateService(func(svc *v1.Service) {
+			svc.Annotations["service.kubernetes.io/topology-aware-hints"] = "disabled"
+		})
+		framework.ExpectNoError(err)
+		// check that hints are disabled
+		err = wait.PollImmediate(3*time.Second, 1*time.Minute, func() (bool, error) {
+			es, err := cs.DiscoveryV1().EndpointSlices(namespace).List(context.TODO(), opts)
+			if err != nil {
+				framework.Logf("Failed to get list EndpointSlice objects: %v", err)
+				// Retry the error
+				return false, nil
+			}
+			for _, endpointSlice := range es.Items {
+				for _, ep := range endpointSlice.Endpoints {
+					if ep.Hints != nil {
+						framework.Logf("EndpointSlice still have topology hints")
+						return false, nil
+					}
+				}
+			}
+			return true, nil
+
+		})
+		framework.ExpectNoError(err)
+
+		err = jig.CheckServiceReachability(svc, execPod)
+		framework.ExpectNoError(err)
+		// get the traffic distribution without Topology Hints
+		zones = map[string]int{}
+		stdout, err = framework.RunHostCmd(execPod.Namespace, execPod.Name, cmd)
+		framework.ExpectNoError(err)
+
+		hostnames = strings.Split(stdout, "\n")
+		failed = 0
+		for _, h := range hostnames {
+			z, ok := podsZones[h]
+			if ok {
+				zones[z]++
+			} else {
+				failed++
+			}
+		}
+		framework.Logf("Connections per zone from %s without topology hints %v Failed %d", nodeZone, zones, failed)
+		// Fail if traffic per zone is not equally distributed [40%,60%]
+		// don't consider the failed connections for the stats
+		// TODO(aojea): based on local testing with KIND revisit percentages.
```

Comment:
This also feels like something that would be quite flaky. It is not really a core feature of topology. Maybe the test could be that when hints are disabled, requests are routed outside of the zone at least sometimes.

Reply:
it is flaky, just wanted to discuss first with you, let's talk in https://github.com/kubernetes/kubernetes/pull/100807